changed method AclRuleL3::validateAddAction() and AclRuleL3::getRedir…

madhanmellanox opened the pull request:
…ectObjectId() to handle all ACL redirect cases

<!--
Please make sure you have read and understood the contribution guildlines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

1. Make sure your commit includes a signature generted with `git commit -s`
2. Make sure your commit title follows the correct format: [component]: description
3. Make sure your commit message contains enough details about the change and related tests
4. Make sure your pull request adds related reviewers, asignees, labels

Please also provide the following information in this pull request:
-->

**What I did**
I made changes to methods AclRuleL3()::validateAddAction() and AclRuleL3::getRedirectObject() in order to accomodate all type of ACL redirect commands.

**Why I did it**
Because, the current code had some assumptions on how the ACL redirect request should be parsed in the ACL orchagent. I removed the assumption.

**How I verified it**
I verified it by testing different kinds of ACL redirect commands on my switch.

**Details if related**

Commit:
changed method AclRuleL3::validateAddAction() and AclRuleL3::getRedirectObjectId() to handle all ACL redirect cases

Commit:
changed method AclRuleL3::validateAddAction() and AclRuleL3::getRedirectObjectId() to handle all ACL redirect cases

stepanblyschak commented on orchagent/aclorch.cpp:
```diff
@@ -809,8 +809,22 @@ bool AclRuleL3::validateAddAction(string attr_name, string _attr_value)
+            // check that we have a colon after redirect rule
+            size_t colon_pos = string(PACKET_ACTION_REDIRECT).length();
+
+            if (attr_value[colon_pos] != ':')
```
This line has UB: if attr_value.length() == colon_pos (e.g.: attr_value == PACKET_ACTION_REDIRECT) it reads behind the allocated string heap (or stack depending on SSO in std library and string size).
Note: c_str() will guaranty to insert '\0' at the end and the above code will become valid.

madhanmellanox replied:
I handled this case in the next commit. Thanks.

Commit:
Handled string handling of attr_value in  AclRuleL3::validateAddAction() method

Commit:
Handled string handling of attr_value in  AclRuleL3::validateAddAction() method

daall committed:
[dvs] Add redis polling interface to dvs fixture (#1228)

* [dvs] Add redis polling interface to dvs fixture
- Remove standalone db fixtures
- Refactor NAT and ACL tests to use new interface

Signed-off-by: Danny Allen <daall@microsoft.com>

daall committed:
[dvs] Add generic polling utility (#1233)

Signed-off-by: Danny Allen <daall@microsoft.com>

wendani committed:
[sub intf] Use m_lag_id to be the parent port object id when parent port is LAG (#1235)

Fix to Azure/sonic-buildimage#4142

Add vs test on lag sub interface based on unconventional lag naming "PortChannel1.20" to fit the interface name within 15 characters

Signed-off-by: Wenda Ni wonda.ni@gmail.com

daall committed:
[portsorch] Enable port-level buffer drop counters (#1237)

Signed-off-by: Danny Allen <daall@microsoft.com>

Commit:
[vnet]: Fix double route installation for BITMAP VNET interface (#1114)

* [vnet]: Fix double route installation for BITMAP VNET interface

qiluo-msft committed:
Add more log message, fix test code (#1239)

qiluo-msft committed:
Update README.md (#1240)

add vlan package for the command line vconfig

daall committed:
[lgtm] Add LGTM checks to vs tests (#1244)

Signed-off-by: Danny Allen <daall@microsoft.com>

qiluo-msft committed:
Refine getDbId() calling to fix build after swss-common change (#1245)

* Refine getDbId() calling to fix build after swss-common change

* Fix mock tests with multiDB support and config

lguohan committed:
configure extra inc/lib directory for build (#1247)

Signed-off-by: Guohan Lu <lguohan@gmail.com>

daall committed:
[dvs] Refactor VLAN tests to use dvslib (#1241)

- Refactor tests to use polling interface
- Refactor redundant testing methods
- Reenable unstable tests

Signed-off-by: Danny Allen <daall@microsoft.com>

daall committed:
[dvs] Re-enable RIF tests (#1249)

Signed-off-by: Danny Allen <daall@microsoft.com>

qiluo-msft committed:
const initializer_list is not a constant expression (#1250)

lguohan committed:
[lgtm]: add lgtm cpp check (#1248)

* [lgtm]: add lgtm cpp check
* add recursive
* add depends

qiluo-msft committed:
Update README.md: improve the style of build badges and add LGTM badges (#1251)

abdosi committed:
[bufferorch] Fixed SAI_BUFFER_PROFILE_ATTR_BUFFER_SIZE to uint64_t (#1255)

as per  SAI 1.5

daall committed:
[dvs] Stabilize sub-port tests (#1254)

Signed-off-by: Danny Allen <daall@microsoft.com>

daall committed:
[dvslib] Improve support for assert logging (#1258)

Signed-off-by: Danny Allen <daall@microsoft.com>

daall committed:
[dvs] Update tests to use correct command to bring up interfaces (#1260)

* [dvs] Update tests to use correct command to bring up interfaces

Signed-off-by: Danny Allen <daall@microsoft.com>

abdosi committed:
Do not set PG to Buffer porfile mapping again if already exist. (#1261)

This will avoid not to call SAI API
set_ingress_priority_group_attribute() if mapping already exist

Signed-off-by: Abhishek Dosi <abdosi@microsoft.com>

qiluo-msft committed:
Fix obvious syntax errors (#1262)

* Fix obvious syntax errors
* Remove unused var

judyjoseph committed:
[Multi-AISC] Pass the asic instance as SAI attribute during switch_create  (#1269)

* SAI 3.7 Supports multiple ASIC instances. This change is needed for orchagent to accept the
instance ID which is fills in HARDWARE_INFO and passes to switch create.

* Update the usage help string with the new option [-i INST_ID]

Commit:
[routeorch] Handle the empty "nexthop" field for backward compatibility (#1263)

* [routeorch] Handle the empty "nexthop" field for backward compatibility

Commit:
[vnet] Verify if BITMAP route exists before creating new one to avoid dublication (#1272)

Signed-off-by: Volodymyr Samotiy <volodymyrs@mellanox.com>

Commit:
[vnet] Set MTU for the VNET bridge RIF in BITMAP implementation (#1271)

Signed-off-by: Volodymyr Samotiy <volodymyrs@mellanox.com>

Commit:
added VS testcases for ACL redirect and priority ACL tests

Conflicts:
	tests/test_acl.py

Commit:
added VS testcases for ACL redirect and priority ACL tests

Conflicts:
	tests/test_acl.py

Conflicts:
	tests/test_acl.py

Commit:
removed test_AclpriorityRule() test case, as it does not belong here

Commit:
Merge branch 'swss_aclredirect' of https://github.com/madhanmellanox/sonic-swss into swss_aclredirect

madhanmellanox commented:
This is an internal PR and Azure PR corresponding to this one got merged, so closing this one.
